rustls/src/builder.rs

```diff
@@ -62,81 +61,70 @@ use crate::versions;
 /// This may be shortened to:
 ///
 /// ```
-/// # use rustls::client_config_builder_with_safe_defaults;
+/// # use rustls::ClientConfig;
 /// # let root_certs = rustls::RootCertStore::empty();
 /// # let trusted_ct_logs = &[];
-/// client_config_builder_with_safe_defaults()
+/// ClientConfig::builder()
+///     .with_safe_defaults()
 ///     .with_root_certificates(root_certs, trusted_ct_logs)
 ///     .with_no_client_auth();
 /// ```
 ///
-///
 /// The types used here fit together like this:
 ///
+/// 1. Call [`ClientConfig::builder()`] or [`ServerConfig::builder()`] to initialize a builder.
 /// 1. You must make a decision on which cipher suites to use, typically
-///    by calling [`ConfigWantsCipherSuites::with_safe_default_cipher_suites()`].
+///    by calling [`ConfigBuilder<S, WantsCipherSuites>::with_safe_default_cipher_suites()`].
 /// 2. Now you must make a decision
-///    on key exchange groups: typically by calling [`ConfigWantsKxGroups::with_safe_default_kx_groups()`].
+///    on key exchange groups: typically by calling
+///    [`ConfigBuilder<S, WantsKxGroups>::with_safe_default_kx_groups()`].
 /// 3. Now you must make
 ///    a decision on which protocol versions to support, typically by calling
-///    [`ConfigWantsVersions::with_safe_default_protocol_versions()`].
-/// 4. You now need to indicate whether to make a [`ServerConfig`] or [`ClientConfig`],
-///    by calling [`ConfigWantsPeerType::for_server()`]
-///    or [`ConfigWantsPeerType::for_client()`] respectively.
-/// 5. Now see [`ConfigWantsServerVerifier`] or [`ConfigWantsClientVerifier`] for further steps.
+///    [`ConfigBuilder<S, WantsVersions>::with_safe_default_protocol_versions()`].
+/// 5. Now see [`ConfigBuilder<ClientConfig, WantsVerifier>`] or
+///    [`ConfigBuilder<ServerConfig, WantsVerifier>`] for further steps.
 ///
 /// [`ServerConfig`]: crate::ServerConfig
 /// [`ClientConfig`]: crate::ClientConfig
-pub fn config_builder() -> ConfigWantsCipherSuites {
-    ConfigWantsCipherSuites {}
+/// [`ClientConfig::builder()`]: crate::ClientConfig::builder()
+/// [`ServerConfig::builder()`]: crate::ServerConfig::builder()
+/// [`ConfigWantsServerVerifier`]: crate::ConfigWantsServerVerifier
+/// [`ConfigWantsClientVerifier`]: crate::ConfigWantsClientVerifier
+#[derive(Clone)]
```

Comment:
should the builder derive `Debug`? if we don't want to implement `Debug` for the various states, it could just be something like this:
```rust
impl<Side: ConfigSide, State> fmt::Debug for ConfigBuilder<Side, State> {
    fn fmt(&self, f: &mut fmt::Formatter<'_> -> fmt::Result {
        f.debug_struct("ConfigBuilder")
            .field("side", format_args!("{}", &std::any::type_name::<Side>()))
            .field("state", format_args!("{}", &std::any::type_name::<State>()))
            .finish()
    }
}
```
or, we could derive `Debug` for the state types, if we want to format their data, and use `type_name` for the `Side` parameter.